sw/device/lib/dif/dif_sram_ctrl.h

```diff
@@ -0,0 +1,215 @@
+ * enable or disable execution from SRAM.
+ *
+ * @param handle A SRAM Controller handle.
+ * @param state[out] Enabled if locked, disabled if unlocked.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_sram_ctrl_result_t dif_sram_ctrl_execution_get_locked(
+    const dif_sram_ctrl_t *handle, dif_sram_ctrl_toggle_t *state);
+
+/**
+ * Reports whether the control register is locked or not.
+ *
+ * If the control register is locked then it is no longer possible to request a
+ * new scrambling key.
+ *
+ * @param handle A SRAM Controller handle.
+ * @param state[out] Enabled if locked, disabled if unlocked.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_sram_ctrl_result_t dif_sram_ctrl_control_get_locked(
+    const dif_sram_ctrl_t *handle, dif_sram_ctrl_toggle_t *state);
+
+/**
+ * Enable or disable execution from SRAM.
+ *
+ * @param handle A SRAM Controller handle.
+ * @param state Enable or disable SRAM execution.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_sram_ctrl_result_t dif_sram_ctrl_execution_set_enabled(
+    const dif_sram_ctrl_t *handle, const dif_sram_ctrl_toggle_t *state);
+
+/**
+ * Report whether execution from SRAM is currently enabled or disabled.
+ *
+ * @param handle A SRAM Controller handle.
+ * @param[out] state State of SRAM execution (enabled or disabled).
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_sram_ctrl_result_t dif_sram_ctrl_execution_get_enabled(
+    const dif_sram_ctrl_t *handle, dif_sram_ctrl_toggle_t *state);
+
+/**
+ * Retrieve the error state from the SRAM controller.
+ *
+ * This function will return `kDifSramCtrlOk` if the error state of the
+ * SRAM controller was successfully fetched regardless of whether the SRAM
+ * controller has experienced an uncorrectable parity error or not. The `error`
+ * argument must be checked to see if a uncorrectable parity error has actually
+ * occurred.
+ *
+ * @param handle A SRAM Controller handle.
+ * @param[out] error Whether an uncorrectable parity error has occurred.
+ * @param[out] address The byte address of the last parity error (optional).
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_sram_ctrl_result_t dif_sram_ctrl_get_error(const dif_sram_ctrl_t *handle,
+                                               bool *error, uintptr_t *address);
+
+/**
+ * Retrieve the current key validity from the SRAM controller.
+ *
+ * If `valid` is set to true then SRAM is scrambled using a valid key from OTP.
+ * Otherwise SRAM is scrambled using a default all-zero key and nonce.
+ *
+ * @param handle A SRAM Controller handle.
+ * @param[out] valid Whether a key has been successfully obtained.
+ * @param[out] seed_valid Whether the scrambling key was derived from a valid
+ * key seed (optional).
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_sram_ctrl_result_t dif_sram_ctrl_get_key_valid(
```

Comment:
Alternative would have been to simply implement `dif_sram_ctrl_get_status` with a `bool *set` as an output parameter. But I do like this and the above routine that combines STATUS and ERROR_ADDRESS information.

Would be interesting to hear what others think.

Reply:
This function now populates a struct as per @mcy's comments. I've kept the error status and scrambling status separate for now, but I am happy to combine them if others think that would be useful.